// freestyle-sandboxes/dist/mastra/index.cjs
'use strict';

var zod = require('zod');
var index = require('../index-DuOpIaWc.cjs');
var index$1 = require('../index.cjs');
require('@hey-api/client-fetch');

var Tool = class {
  id;
  description;
  inputSchema;
  outputSchema;
  execute;
  mastra;
  constructor(opts) {
    this.id = opts.id;
    this.description = opts.description;
    this.inputSchema = opts.inputSchema;
    this.outputSchema = opts.outputSchema;
    this.execute = opts.execute;
    this.mastra = opts.mastra;
  }
};
function createTool(opts) {
  return new Tool(opts);
}

const executeTool = (config) => {
  const description = index.executeCodeDescription(
    Object.keys(config.envVars ?? {}).join(", "),
    Object.keys(config.nodeModules ?? {}).join(", ")
  );
  const client = new index$1.FreestyleSandboxes({
    apiKey: config.apiKey
  });
  return createTool({
    id: "Execute a TypeScript or JavaScript Script",
    description,
    execute: async ({ context: { script } }) => {
      return await client.executeScript(script, config);
    },
    inputSchema: index.executeCodeSchema,
    outputSchema: zod.z.object({
      logs: zod.z.array(
        zod.z.object({
          message: zod.z.string(),
          type: zod.z.string()
        })
      ),
      result: zod.z.unknown()
    })
  });
};

exports.executeTool = executeTool;
