// freestyle-sandboxes/dist/react/dev-server/index.cjs
'use strict';

var reactQuery = require('@tanstack/react-query');
var React = require('react');

const queryClient = new reactQuery.QueryClient();
function DefaultLoadingComponent({
  installCommandRunning
}) {
  let loadingText = "Starting container...";
  if (installCommandRunning) {
    loadingText = "Installing dependencies...";
  } else {
    loadingText = "Starting dev server...";
  }
  return /* @__PURE__ */ React.createElement(
    "div",
    {
      style: {
        display: "flex",
        flexDirection: "column",
        alignItems: "center",
        justifyContent: "center",
        height: "100%"
      }
    },
    loadingText
  );
}
function FreestyleDevServer({
  loadingComponent,
  actions,
  repoId
}) {
  return /* @__PURE__ */ React.createElement(reactQuery.QueryClientProvider, { client: queryClient }, /* @__PURE__ */ React.createElement(
    FreestyleDevServerInner,
    {
      loadingComponent: loadingComponent ?? DefaultLoadingComponent,
      repoId,
      actions
    }
  ));
}
function FreestyleDevServerInner({
  repoId,
  loadingComponent,
  actions: { requestDevServer }
}) {
  const { data, isLoading } = reactQuery.useQuery({
    queryKey: ["dev-server", repoId],
    queryFn: async () => await requestDevServer({ repoId }),
    refetchInterval: 1e3
  });
  const ref = React.useRef(null);
  const [wasLoaded, setWasLoaded] = React.useState(false);
  const [iframeLoaded, setIframeLoaded] = React.useState(false);
  React.useMemo(() => {
    if (data?.devCommandRunning) {
      setWasLoaded(true);
    }
  }, [isLoading, data?.devCommandRunning]);
  React.useEffect(() => {
    function loadHandle() {
      setIframeLoaded(true);
    }
    ref.current?.addEventListener("load", loadHandle);
    return () => {
      ref.current?.removeEventListener("load", loadHandle);
    };
  }, [ref]);
  if (isLoading) {
    return loadingComponent({
      devCommandRunning: false,
      installCommandRunning: false,
      serverStarting: true,
      iframeLoading: false
    });
  }
  if (!data?.devCommandRunning && !wasLoaded) {
    return loadingComponent({
      devCommandRunning: data?.devCommandRunning ?? false,
      installCommandRunning: data?.installCommandRunning ?? false,
      serverStarting: false,
      iframeLoading: false
    });
  }
  return /* @__PURE__ */ React.createElement(
    "div",
    {
      style: {
        display: "grid",
        gridTemplateRows: "1fr",
        gridTemplateColumns: "1fr",
        width: "100%",
        height: "100%"
      }
    },
    /* @__PURE__ */ React.createElement(
      "div",
      {
        style: {
          width: "100%",
          height: "100%",
          border: "none",
          gridColumn: "1 / -1",
          gridRow: "1 / -1",
          visibility: iframeLoaded ? "hidden" : "visible"
        }
      },
      loadingComponent({
        devCommandRunning: data?.devCommandRunning ?? false,
        installCommandRunning: data?.installCommandRunning ?? false,
        serverStarting: false,
        iframeLoading: true
      })
    ),
    /* @__PURE__ */ React.createElement(
      "iframe",
      {
        ref,
        sandbox: "allow-scripts allow-same-origin allow-forms",
        src: data.ephemeralUrl,
        style: {
          width: "100%",
          height: "100%",
          border: "none",
          gridColumn: "1 / -1",
          gridRow: "1 / -1"
        }
      }
    )
  );
}

exports.DefaultLoadingComponent = DefaultLoadingComponent;
exports.FreestyleDevServer = FreestyleDevServer;
